### PR TITLE
Fix bool support in reverse engineering and correct type mappings

### LINKED ISSUE #654: Can't read TINYINT(1) as bool
### Steps to reproduce
```
using System.Threading.Tasks;
using FluentAssertions;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Diagnostics;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Console;
using NUnit.Framework;

namespace Tests
{
    [TestFixture, Explicit]
    class MySqlCustomModelTests
    {
        static readonly LoggerFactory _loggerFactory
            = new LoggerFactory(new[] { new ConsoleLoggerProvider((category, logLevel) => true, false) });
        DbContextOptions<Context> _options;

        Context GetContext()
        {
            return new Context(_options);
        }

        class Context : DbContext
        {
            public Context(DbContextOptions<Context> options) : base(options) { }

            protected override void OnModelCreating(ModelBuilder modelBuilder)
            {
                base.OnModelCreating(modelBuilder);

                modelBuilder.Entity<Bool>(entity =>
                {
                    entity.ToTable("bools");
                    entity.Property(e => e.BoolValue).HasColumnType("tinyint(1)");
                });
            }

            public DbSet<Bool> Bools { get; set; }
        }

        class Bool
        {
            public int Id { get; set; }
            public bool BoolValue { get; set; }
        }

        [SetUp]
        protected void BeforeEach()
        {
            _options = new DbContextOptionsBuilder<Context>()
                .UseLoggerFactory(_loggerFactory)
                .UseMySql("Server=localhost;database=createtest;uid=user;pwd=password")
                .ConfigureWarnings(warnings => warnings.Throw(CoreEventId.IncludeIgnoredWarning))
                .Options;

            using (var db = GetContext())
            {
                db.Database.EnsureDeleted();
                db.Database.EnsureCreated();
            }
        }

        [Test]
        public async Task CanReadBools()
        {
            using (var db = GetContext())
            {
                db.Bools.Add(new Bool { BoolValue = false });
                db.Bools.Add(new Bool { BoolValue = true });
                await db.SaveChangesAsync();
            }

            using (var db = GetContext())
            {
                var list = await db.Bools.ToListAsync();
                list[0].BoolValue.Should().BeFalse();
                list[1].BoolValue.Should().BeTrue();
            }
        }
    }
}
```
### The issue

Can't read TINYINT(1) mapped to bool
It's not MySqlConnector issue, it reads fine with reader.GetBoolean(1)

```
Exception: 
System.InvalidOperationException : An exception occurred while reading a database value for property 'Bool.BoolValue'. The expected type was 'System.Boolean' but the actual value was of type 'System.Boolean'.
  ----> System.InvalidCastException : Unable to cast object of type 'System.Boolean' to type 'System.SByte'.
   at Microsoft.EntityFrameworkCore.Metadata.Internal.EntityMaterializerSource.ThrowReadValueException[TValue](Exception exception, Object value, IPropertyBase property)
   at lambda_method(Closure , DbDataReader )
   at Microsoft.EntityFrameworkCore.Storage.Internal.TypedRelationalValueBufferFactory.Create(DbDataReader dataReader)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.BufferlessMoveNext(DbContext _, Boolean buffer, CancellationToken cancellationToken)
   at Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.MoveNextCore(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.AsyncIterator`1.MoveNext(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.Aggregate_[TSource,TAccumulate,TResult](IAsyncEnumerable`1 source, TAccumulate seed, Func`3 accumulator, Func`2 resultSelector, CancellationToken cancellationToken)
   at Tests.MySqlCustomModelTests.CanReadBools() in C:\Dev\Tests\MySqlCustomModelTests.cs:line 75
   at NUnit.Framework.Internal.AsyncInvocationRegion.AsyncTaskInvocationRegion.WaitForPendingOperationsToComplete(Object invocationResult) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\AsyncInvocationRegion.cs:line 113
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunAsyncTestMethod(TestExecutionContext context) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 96
--InvalidCastException
   at MySql.Data.MySqlClient.MySqlDataReader.GetFieldValue[T](Int32 ordinal) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 286
   at lambda_method(Closure , DbDataReader )
```

### Further technical details

MySQL version: 8.0.11
Operating system: Win 7
Pomelo.EntityFrameworkCore.MySql version: 2.1.1

### LINKED ISSUE #656: Respect TreatTinyAsBoolean connection string option
### Background
Based mostly on [this article](https://www.xaprb.com/blog/2006/04/11/bit-values-in-mysql/) and the fact that bit(1) in MySql doesn't provide any space saving over tinyint(1) and how weirdly it's represented in various MySql tools I made a decision to use tinyint(1) for boolean values in my project. The article is from 2006, so some things mentioned there may have improved since, but still I feel that tinyint is a safer option - wouldn't want to run into some obscure issues in prodution.

### The issue
I use db first and scaffold models when I add tables, but I also generate test dbs off the models during some tests. With 2.0.1 I had to change sbyte to bool in scaffolded models. 2.1.1 [broke this at runtime](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/654). Current workaround is to use TreatTinyAsBoolean=false. I believe this is incorrect and having that option as true (default) should properly map tinyint(1) to bool when both scaffolding and generating tables (and work at runtime of course).

Yes I know that technically tinyint(1) can hold the same range of values as tinyint(4), but in this case I think it's irrelevant, as TreatTinyAsBoolean, as the [implementation in the underlying provider intends](https://github.com/mysql-net/MySqlConnector/issues/141), is specifically about treating (only) tinyint(1) (and should not affect other tinyint display sizes). And please don't bring ZeroFill into this :)

### Further technical details

MySQL version: 8.0.11
Operating system: Win 7 x64
Pomelo.EntityFrameworkCore.MySql version: 2.1.1

### LINKED ISSUE #866: Code generated by scaffolder leads to: Unable to cast object of type 'System.Boolean' to type 'System.SByte'
Hi

I use dotnet core MySql package 3.0.0-rc1.final

In DB I have columns as tinyint(1) and they are scaffolded as sbyte, but when I make select i got this error: Unable to cast object of type 'System.Boolean' to type 'System.SByte'.'

Is it possible to fix it? or do I something wrong?

Thanks

### PR DESCRIPTION
**Pull request by lauxjpn**
See #866 for an extensive discussion about the broken scaffolding and bool support in general.

Let's assume the following table:

```sql
CREATE TABLE `ScaffoldMe` (
  `ScaffoldMeId` int(11) NOT NULL AUTO_INCREMENT,
  `NonNullableTinyInt1` tinyint(1) NOT NULL,
  `NonNullableTinyInt2` tinyint(2) NOT NULL,
  `NonNullableBit1` bit(1) NOT NULL,
  `NonNullableBit2` bit(2) NOT NULL,
  `NullableTinyInt1` tinyint(1) DEFAULT NULL,
  `NullableTinyInt2` tinyint(2) DEFAULT NULL,
  `NullableBit1` bit(1) DEFAULT NULL,
  `NullableBit2` bit(2) DEFAULT NULL,
  PRIMARY KEY (`ScaffoldMeId`)
) ENGINE=InnoDB;
```

Before this PR, the following code was generated by the scaffolding process while completely ignoring the `TreatTinyAsBoolean` connection string option and lead to a broken model and model definition:

```c#
using System;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata;
using System.Collections.Generic;

namespace IssueConsoleTemplate.Models
{
    public partial class ScaffoldMe
    {
        public int ScaffoldMeId { get; set; }
        public sbyte NonNullableTinyInt1 { get; set; }
        public sbyte NonNullableTinyInt2 { get; set; }
        public bool NonNullableBit1 { get; set; }
        public bool NonNullableBit2 { get; set; }
        public sbyte? NullableTinyInt1 { get; set; }
        public sbyte? NullableTinyInt2 { get; set; }
        public bool? NullableBit1 { get; set; }
        public bool? NullableBit2 { get; set; }
    }

    public partial class Issue866Context : DbContext
    {
        public Issue866Context()
        {
        }

        public Issue866Context(DbContextOptions<Issue866Context> options)
            : base(options)
        {
        }

        public virtual DbSet<ScaffoldMe> ScaffoldMe { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            if (!optionsBuilder.IsConfigured)
            {
                optionsBuilder.UseMySql(
"server=127.0.0.1;user id=root;password=;port=3308;database=Issue866;TreatTinyAsBoolean=true");
            }
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<ScaffoldMe>(entity =>
            {
                entity.Property(e => e.ScaffoldMeId).HasColumnType("int(11)");
                entity.Property(e => e.NonNullableBit1).HasColumnType("bit(1)");
                entity.Property(e => e.NonNullableBit2).HasColumnType("bit(2)");
                entity.Property(e => e.NonNullableTinyInt1).HasColumnType("tinyint(1)");
                entity.Property(e => e.NonNullableTinyInt2).HasColumnType("tinyint(2)");
                entity.Property(e => e.NullableBit1).HasColumnType("bit(1)");
                entity.Property(e => e.NullableBit2).HasColumnType("bit(2)");
                entity.Property(e => e.NullableTinyInt1).HasColumnType("tinyint(1)");
                entity.Property(e => e.NullableTinyInt2).HasColumnType("tinyint(2)");
            });

            OnModelCreatingPartial(modelBuilder);
        }

        partial void OnModelCreatingPartial(ModelBuilder modelBuilder);
    }
}
```

With this PR and `TreatTinyAsBoolean=true` (the default), the following working code will be generated, that maps `tinyint(1)` to `System.Boolean` and is the preferred option to move forward:

```c#
using System;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata;
using System.Collections.Generic;

namespace IssueConsoleTemplate.Models
{
    public partial class ScaffoldMe
    {
        public int ScaffoldMeId { get; set; }
        public bool NonNullableTinyInt1 { get; set; }
        public sbyte NonNullableTinyInt2 { get; set; }
        public ulong NonNullableBit1 { get; set; }
        public ulong NonNullableBit2 { get; set; }
        public bool? NullableTinyInt1 { get; set; }
        public sbyte? NullableTinyInt2 { get; set; }
        public ulong? NullableBit1 { get; set; }
        public ulong? NullableBit2 { get; set; }
    }

    public partial class Issue866Context : DbContext
    {
        public Issue866Context()
        {
        }

        public Issue866Context(DbContextOptions<Issue866Context> options)
            : base(options)
        {
        }

        public virtual DbSet<ScaffoldMe> ScaffoldMe { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            if (!optionsBuilder.IsConfigured)
            {
                optionsBuilder.UseMySql(
"server=127.0.0.1;user id=root;password=;port=3308;database=Issue866;TreatTinyAsBoolean=true");
            }
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<ScaffoldMe>(entity =>
            {
                entity.Property(e => e.ScaffoldMeId).HasColumnType("int(11)");
                entity.Property(e => e.NonNullableBit1).HasColumnType("bit(1)");
                entity.Property(e => e.NonNullableBit2).HasColumnType("bit(2)");
                entity.Property(e => e.NonNullableTinyInt2).HasColumnType("tinyint(2)");
                entity.Property(e => e.NullableBit1).HasColumnType("bit(1)");
                entity.Property(e => e.NullableBit2).HasColumnType("bit(2)");
                entity.Property(e => e.NullableTinyInt2).HasColumnType("tinyint(2)");
            });

            OnModelCreatingPartial(modelBuilder);
        }

        partial void OnModelCreatingPartial(ModelBuilder modelBuilder);
    }
}
```

With this PR and `TreatTinyAsBoolean=false`, the following working code will be generated, that maps `bit(1)` to `System.Boolean` and is mainly backwards compatible with previous Pomelo versions:

```c#
using System;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata;
using System.Collections.Generic;

namespace IssueConsoleTemplate.Models
{
    public partial class ScaffoldMe
    {
        public int ScaffoldMeId { get; set; }
        public sbyte NonNullableTinyInt1 { get; set; }
        public sbyte NonNullableTinyInt2 { get; set; }
        public bool NonNullableBit1 { get; set; }
        public ulong NonNullableBit2 { get; set; }
        public sbyte? NullableTinyInt1 { get; set; }
        public sbyte? NullableTinyInt2 { get; set; }
        public bool? NullableBit1 { get; set; }
        public ulong? NullableBit2 { get; set; }
    }

    public partial class Issue866Context : DbContext
    {
        public Issue866Context()
        {
        }

        public Issue866Context(DbContextOptions<Issue866Context> options)
            : base(options)
        {
        }

        public virtual DbSet<ScaffoldMe> ScaffoldMe { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            if (!optionsBuilder.IsConfigured)
            {
                optionsBuilder.UseMySql(
"server=127.0.0.1;user id=root;password=;port=3308;database=Issue866;TreatTinyAsBoolean=false");
            }
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<ScaffoldMe>(entity =>
            {
                entity.Property(e => e.ScaffoldMeId).HasColumnType("int(11)");
                entity.Property(e => e.NonNullableBit2).HasColumnType("bit(2)");
                entity.Property(e => e.NonNullableTinyInt1).HasColumnType("tinyint(1)");
                entity.Property(e => e.NonNullableTinyInt2).HasColumnType("tinyint(2)");
                entity.Property(e => e.NullableBit2).HasColumnType("bit(2)");
                entity.Property(e => e.NullableTinyInt1).HasColumnType("tinyint(1)");
                entity.Property(e => e.NullableTinyInt2).HasColumnType("tinyint(2)");
            });

            OnModelCreatingPartial(modelBuilder);
        }

        partial void OnModelCreatingPartial(ModelBuilder modelBuilder);
    }
}
```

This PR correctly supports the `TreatTinyAsBoolean` connection string option when reverse engineering and in the provider in general and fixes other type mapping issues.

Any review or actual production code test of this PR is highly appreciated.

Fixes #866 
Fixes #656 
Fixes #654

### CODE REVIEW

**Comment by bgrainger** [comment]
Does this work by looking at the SQL (as a string) or the actual field types? In MySQL the `BOOL` and `BOOLEAN` keywords "[are synonyms for `TINYINT(1)`](https://dev.mysql.com/doc/refman/8.0/en/numeric-type-overview.html)". Thus, if this is looking for the text `tinyint(1)` it should also check for `bool` and `boolean`. (If it uses field types from the DB then everything should be fine.)

File: src/EFCore.MySql/Storage/Internal/MySqlTypeMappingSource.cs [hunk 1]
```diff
@@ -298,6 +324,16 @@ private RelationalTypeMapping FindRawMapping(RelationalTypeMappingInfo mappingIn
...
+                    else if (storeTypeName.Equals("tinyint(1)", StringComparison.OrdinalIgnoreCase)
+                             && _options.ConnectionSettings.TreatTinyAsBoolean)
+                    {
+                        return _tinyint1;
+                    }
+                    else if (storeTypeName.Equals("bit(1)", StringComparison.OrdinalIgnoreCase)
+                             && !_options.ConnectionSettings.TreatTinyAsBoolean)
+                    {
+                        return _bit1;
+                    }
```
Comment (twsI) [inline src/EFCore.MySql/Storage/Internal/MySqlTypeMappingSource.cs]: Why not use the StoreType of the _tinyint1 and _bit1 field?

**Comment by lauxjpn** [comment]
@bgrainger 
An actual lookup will only happen when scaffolding/reverse engineering from an existing database by using `MySqlDatabaseModelFactory` to execute a `SHOW COLUMNS FROM` statement.
According to the [MySQL docs](https://dev.mysql.com/doc/refman/8.0/en/other-vendor-data-types.html), this cannot contain any compatibility types.

It is however possible, that someone is manually creating a model definition using compatibility types like this:

```c#
modelBuilder.Property(e => e.SomeProperty).HasColumnType("CHARACTER VARYING(127)");
```

There is no mapping support for types from other vendors implemented in Pomelo.

It's a question of whether we should encourage the usage of types in model definitions, that exist only for compatibility reasons.
If we do want to support such types, I can push another quick commit that implements them.

File: src/EFCore.MySql/Storage/Internal/MySqlTypeMappingSource.cs (hunk 1 above)
Comment (lauxjpn) [reply src/EFCore.MySql/Storage/Internal/MySqlTypeMappingSource.cs]: That's a valid point. I replaced all of them with field references now.

Comment (lauxjpn) [reply src/EFCore.MySql/Storage/Internal/MySqlTypeMappingSource.cs]: Thinking about this a bit more, it should be possible to move most of the dynamic type checking out of `FindRawMapping()` and into `Initialize()`.

**Comment by bgrainger** [comment]
> It's a question of whether we should encourage the usage of types in model definitions, that exist only for compatibility reasons.

The current approach (handling what is returned from `SHOW COLUMNS FROM`) sounds best to me.

**Pull request merged**